// aws-crt/dist/index.js
"use strict";
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrtError = exports.resource_safety = exports.platform = exports.iot = exports.auth = exports.crypto = exports.http = exports.mqtt = exports.io = exports.crt = void 0;
/**
 * @packageDocumentation
 * @module crt
 */
// This is the entry point for the AWS CRT nodejs native libraries
/* common libs */
const platform = __importStar(require("./common/platform"));
exports.platform = platform;
const resource_safety = __importStar(require("./common/resource_safety"));
exports.resource_safety = resource_safety;
/* node specific libs */
const crt = __importStar(require("./native/crt"));
exports.crt = crt;
const io = __importStar(require("./native/io"));
exports.io = io;
const mqtt = __importStar(require("./native/mqtt"));
exports.mqtt = mqtt;
const http = __importStar(require("./native/http"));
exports.http = http;
const crypto = __importStar(require("./native/crypto"));
exports.crypto = crypto;
const auth = __importStar(require("./native/auth"));
exports.auth = auth;
const iot = __importStar(require("./native/aws_iot"));
exports.iot = iot;
const error_1 = require("./native/error");
Object.defineProperty(exports, "CrtError", { enumerable: true, get: function () { return error_1.CrtError; } });
